OnProfileUpdate/handler.ts

```diff
@@ -0,0 +1,162 @@
+import { DataTableProfileEmailsRepository } from "@pagopa/io-functions-commons/dist/src/utils/unique_email_enforcement/storage";
+import { ProfileEmail } from "@pagopa/io-functions-commons/dist/src/utils/unique_email_enforcement";
+import {
+  ProfileModel,
+  Profile
+} from "@pagopa/io-functions-commons/dist/src/models/profile";
+import { generateVersionedModelId } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model_versioned";
+import { CosmosErrors } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
+import { Logger } from "@azure/functions";
+import { FiscalCode } from "../generated/backend/FiscalCode";
+
+const ProfileDocument = t.intersection([
+  ProfileEmail,
+  t.type({
+    isEmailValidated: t.boolean,
+    version: NonNegativeInteger
+  })
+]);
+
+type ProfileDocument = t.TypeOf<typeof ProfileDocument>;
+
+interface IDependencies {
+  readonly dataTableProfileEmailsRepository: DataTableProfileEmailsRepository;
+  readonly profileModel: ProfileModel;
+  readonly logger: { readonly error: Logger["error"] };
+}
+
+const getPreviousProfile = (
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+) => (dep: IDependencies): TE.TaskEither<CosmosErrors, O.Option<Profile>> =>
+  pipe(
+    version - 1,
+    NonNegativeInteger.decode,
+    E.fold(
+      () => TE.right(O.none),
+      previousVersion =>
+        pipe(
+          generateVersionedModelId<Profile, "fiscalCode">(
+            fiscalCode,
+            previousVersion
+          ),
+          id => dep.profileModel.find([id, fiscalCode])
+        )
+    )
+  );
+
+const deleteProfileEmail = (profileEmail: ProfileEmail) => ({
+  dataTableProfileEmailsRepository
+}: IDependencies): TE.TaskEither<Error, void> =>
+  TE.tryCatch(
+    () => dataTableProfileEmailsRepository.delete(profileEmail),
+    error =>
+      error instanceof Error
+        ? error
+        : new Error("error deleting ProfileEmail from table storage")
+  );
+
+const insertProfileEmail = (profileEmail: ProfileEmail) => ({
+  dataTableProfileEmailsRepository
+}: IDependencies): TE.TaskEither<Error, void> =>
+  TE.tryCatch(
+    () => dataTableProfileEmailsRepository.insert(profileEmail),
+    error =>
+      error instanceof Error
+        ? error
+        : new Error("error inserting ProfileEmail from table storage")
+  );
+
+/*
+If the current email is validated but the previous email was not validated => it inserts the new email into profileEmails
+If the current email is not validated but the previous email was validated => it deletes the previous email from profileEmails
+*/
+const handlePositiveVersion = ({
+  email,
+  fiscalCode,
+  isEmailValidated,
+  version
+}: ProfileDocument): RTE.ReaderTaskEither<
+  IDependencies,
+  Error | CosmosErrors,
+  void
+> =>
+  pipe(
+    getPreviousProfile(fiscalCode, version),
+    RTE.chainW(
+      flow(
+        O.foldW(
+          () => RTE.right<IDependencies, Error, void>(void 0),
```

Comment:
This case should never happen as we should always find a previous version, when the current version is higher than zero. We should add a custom event to track this scenario.

Reply:
If it should never happen, why create a custom event?

Reply:
Should never happen doesn't mean it never will. It's better to add a custom event to check this scenario.